tests/unit/metadata/test_multi_table.py

```diff
@@ -17,6 +20,386 @@ def test___init__(self):
+                parent_table_name,
+                parent_primary_key,
+                child_table_name,
+                child_foreign_key
+            )
+
+    def test__validate_circular_relationships(self):
+        """Test the ``_validate_circular_relationships`` method of ``MultiTableMetadata``.
+
+        Validate that the ``_validate_circular_relationships`` updates the ``errors`` list
+        with the detected circular relationship.
+
+        Setup:
+            - ``child_map``, mapping of parent tables with their child tables.
+
+        Input:
+            - ``parent`` the name of the table that we want to add a relationship for.
+            - ``child_map``, the mapping created previously
+            - ``errors``, the list of errors that has to be updated.
+
+        Side Effects:
+            - The input list has been updated with the tables detected to cause circular
+              relationship.
+        """
+        # Setup
+        child_map = {
+            'users': {'sessions', 'transactions'},
+            'sessions': {'users', 'transactions'}
+        }
+        parent = 'users'
+        errors = []
+
+        # Run
+        MultiTableMetadata()._validate_circular_relationships(
+            parent, child_map=child_map, errors=errors)
+
+        # Assert
+        assert errors == ['users']
+
+    def test_add_relationship_raises_value_error(self):
+        """Test the ``add_relationship`` method of ``MultiTableMetadata``.
+
+        Test that when a circular relationship occurs with the new relationship,
+        a ``ValueError`` is being raised.
+
+        Setup:
+            - Instance of ``MultiTableMetadata``.
+            - Mock of ``parent_table`` simulating a ``SingleTableMetadata``.
+            - Update ``_tables``.
+
+        Input:
+            - ``parent_table_name`` string representing the parent table name.
+            - ``child_table_name`` string representing the child table name.
+            - ``parent_primary_key`` string representing the ``primary_key`` of the parent.
+            - ``child_foreign_key`` string representing the ``foreing_Key`` of the child table.
+
+        Side Effects:
+            - ``ValueError`` is raised.
+        """
+        # Setup
+        instance = MultiTableMetadata()
+        parent_table = Mock()
+        parent_table._primary_key = 'user_id'
+
+        parent_table._columns = {
+            'user_id': {'sdtype': 'numerical'},
+            'session_id': {'sdtype': 'numerical'},
+            'transaction': {'sdtype': 'numerical'}
+        }
+        instance._tables = {
+            'users': parent_table,
+            'sessions': Mock(),
+            'transactions': Mock()
+        }
+        instance._relationships = [
+            {'parent_table_name': 'users', 'child_table_name': 'sessions'},
+            {'parent_table_name': 'sessions', 'child_table_name': 'users'}
+        ]
+
+        # Run / Assert
+        # TODO fix match error message
```

Comment:
?

Reply:
This is a `set` and always comes in 2 pairs, so I will fix with the `SingleTableMetadata` improvement that I'm doing